text/0000-update-json-cnb-registry.md

```diff
@@ -0,0 +1,90 @@
+
+# Summary
+[summary]: #summary
+
+This proposal corrects some mistakes and inefficiencies in the JSON structure proposed by [RFC-0022](https://github.com/buildpacks/rfcs/blob/master/text/0022-client-side-buildpack-registry.md).
+
+# Motivation
+[motivation]: #motivation
+
+During implementation, we found that the JSON structure described in [RFC-0022](https://github.com/buildpacks/rfcs/blob/master/text/0022-client-side-buildpack-registry.md) was both incorrect and redundant.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Each file representing a buildpack in the buildpack registry index will contain JSON. The file will contain multiple entries representing each version of the buildpack split by a newline. However, each version entry will be minimized (stored on a single line), and multiple entries in the same file will not have a comma at the end (i.e. there will be no list/array using `[]`). Thus, each index file *will contain invalid JSON*. But this allows for smaller diffs and simpler appends. This follows the precedent set by [crates.io-index](https://github.com/rust-lang/crates.io-index), which strikes a balance between human redable, easy to parse, and minimizing the diffs for new updates.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+An entry will have the following structure:
+
+```
+{
+  "ns" : "<string>",
+  "name": "<string>",
+  "version" : "<string",
+  "yanked" : <boolean>,
+  "addr" : "<string>",
+}
+```
+
+The fields are defined as follows:
+
+* `ns` - can represent a set or organization of buildpacks.
+* `name` - an identifier that must be unique within a namespace.
+* `version` - the version of the buildpack (must match the version in the `buildpack.toml` of the buildpack)
+* `yanked` - whether or not the buildpack has been removed from the registry
+* `addr` - the address of the image stored in a Docker Registry (ex. `"docker.io/jkutner/lua@sha256:abc123"`). The image reference must use `@digest` and not use `:tag`.
+
+An example of what this may look like for a single buildpack file:
+
+```
+{"ns":"heroku","name":"ruby","version":"0.1.0","yanked":false,"addr":"docker.io/hone/ruby-buildpack@sha256:a9d9038c0cdbb9f3b024aaf4b8ae4f894ea8288ad0c3bf057d1157c74601b906"}
+{"ns":"heroku","name":"ruby","version":"0.2.0","yanked":false,"addr":"docker.io/hone/ruby-buildpack@sha256:2560f05307e8de9d830f144d09556e19dd1eb7d928aee900ed02208ae9727e7a"}
+{"ns":"heroku","name":"ruby","version":"0.2.1","yanked":false,"addr":"docker.io/hone/ruby-buildpack@sha256:74eb48882e835d8767f62940d453eb96ed2737de3a16573881dcea7dea769df7"}
+{"ns":"heroku","name":"ruby","version":"0.3.0","yanked":false,"addr":"docker.io/hone/ruby-buildpack@sha256:8c27fe111c11b722081701dfed3bd55e039b9ce92865473cf4cdfa918071c566"}
+```
+
+*Note:* id is the combination of two fields, `ns` and `name`. The `/` will be replaced by a `_` in the filename. The directories will use the first four characaters of the buildpacks `name`, split into two two character dirs. For example:
```

Comment:
Based on some of the other comments in https://github.com/buildpacks/rfcs/pull/35.

Moving to using the buildpack name for the directory structure seems like a good idea for improving the sharding. It does seem like discoverability/troubleshooting would be impacted. I wonder if the original alternative proposal named "Namespace as top-level dirs" would suite both requirements better.

```
heroku
└── ja
    ├── va
    │   ├── heroku_java
    │   └── heroku_javascript
    └── co
        └── heroku_jacoco
```

Reply:
Adding a more detailed examples:

##### without ns

```
ja
├── va
│   ├── example_java
│   ├── example_javascript
│   ├── heroku_java
│   └── heroku_javascript
└── co
    └── heroku_jacoco
```

##### with ns

```
heroku
└── ja
    ├── va
    │   ├── heroku_java
    │   └── heroku_javascript
    └── co
        └── heroku_jacoco
example
└── ja
    └── va
        ├── example_java
        └── example_javascript

```

Reply:
Based on the context/history provided in the working group meeting I would consider this conversation resolved.